File: bilet.erl
-module(bilet).
-behaviour(world).
-export([start/0, hash/1, br/1, is_dead/1, is_target/1 ]).

-export([cnt/0]).

%%кол-во найденных результатов
cnt() -> cnt(0).

cnt(N) ->
    receive
    inc -> cnt(N+1);
    {cnt,Pid} ->
        Pid ! N,
        cnt(N)
    end.

%%ХЭШ мира. используем стандартный
hash(Wrld) ->
	world:std_hash(Wrld).
	
%%Мир достиг цели?
is_target(Wrld) ->
    {_St_d,_En_d,X1,X2,X3,X4,X5,X6} = Wrld,
    % if ((X1+X2+X3)==(X4+X5+X6)) ->
	% 	global:send("cnt_srv", inc),
	% 	global:send("cnt_srv", {cnt, self()}),
    %     receive
    %         X -> ok
    %     end,
    %     io:format("~w ~w ~w ~w ~w ~w (~w)~n", [X1,X2,X3,X4,X5,X6, X]);
    %     true -> ok
    % end,
    ((X1+X2+X3)==(X4+X5+X6)).
	
%%Мир - тупиковая ветвь
is_dead(Wrld) ->
    {St_d,En_d,_X1,_X2,_X3,_X4,_X5,_X6} = Wrld,
    (St_d == En_d).
	
%%Ветвление мира
br(Wrld) ->
    {St_d,En_d,X1,X2,X3,X4,X5,X6} = Wrld,
    THalf = round((St_d + En_d) / 2),
    if
        (St_d == En_d) ->
            [];
        ((En_d - St_d) == 1) -> 
            XX6 = En_d rem 10,
            XX5 = trunc((En_d rem 100)/10),
            XX4 = trunc((En_d rem 1000)/100),
            XX3 = trunc((En_d rem 10000)/1000),
            XX2 = trunc((En_d rem 100000)/10000),
            XX1 = trunc((En_d rem 1000000)/100000),
            [{St_d,St_d,XX1,XX2,XX3,XX4,XX5,XX6}] ++
            [{En_d,En_d,XX1,XX2,XX3,XX4,XX5,XX6}];
        true ->
            br({St_d,THalf,X1,X2,X3,X4,X5,X6}) ++
            br({THalf,En_d,X1,X2,X3,X4,X5,X6})
    end.

start() ->
	io:format("start~n", []),
    PID_cnt_srv = spawn(?MODULE, cnt, []),	
	global:register_name("cnt_srv", PID_cnt_srv),
	
	world:start(?MODULE,[{0,999999,0,0,0,0,0,0}]),
	ok.

File: bilet.ex
defmodule Bilet do
  require Logger
  @behaviour World

  # хранилище найденных результатов
  def cnt(), do: cnt(%{})
  def cnt(results) do
    receive do
      :stop -> :ok
      {{:add, result}, pid} ->
        new_results = Map.merge(results, %{result => :ok})
        l1 = length(Map.keys(new_results))
        l2 = length(Map.keys(results))
        if (l1 != l2) do
          Logger.info("Target #{inspect(result)}")
        end
        # send(pid, length(Map.keys(new_results)))
        cnt(new_results)
      _ -> :ok
    end
  end

  # ХЭШ мира. используем стандартный
  def hash(wrld), do:
    World.std_hash(wrld)

  # Мир достиг цели?
  def is_target({_st_d,_en_d,x1,x2,x3,x4,x5,x6} = _wrld) do
    if ((x1+x2+x3)==(x4+x5+x6)) do
      # Logger.info("Target #{inspect({x1,x2,x3,x4,x5,x6})}")
      :global.send('res_srv', {{:add, {x1,x2,x3,x4,x5,x6}}, self()})
      # receive do
      #   xxx ->
      #     Logger.info("Target #{inspect({x1,x2,x3,x4,x5,x6})} #{xxx}")
      # end

    end
    ((x1+x2+x3)==(x4+x5+x6))
  end

  # Мир - тупиковая ветвь
  def is_dead({st_d,en_d,_x1,_x2,_x3,_x4,_x5,_x6} = _wrld) do
    st_d == en_d
  end

  # Ветвление мира
  def br({st_d,en_d,x1,x2,x3,x4,x5,x6} = _wrld) do
    # Logger.info("br #{inspect({st_d,en_d})}")
    t_half = round((st_d + en_d) / 2)
    cond do
      (st_d == en_d) -> []
      ((en_d - st_d) == 1) ->
        xx6 = rem(en_d, 10)
        xx5 = trunc(rem(en_d, 100)/10)
        xx4 = trunc(rem(en_d, 1000)/100)
        xx3 = trunc(rem(en_d, 10000)/1000)
        xx2 = trunc(rem(en_d, 100000)/10000)
        xx1 = trunc(rem(en_d, 1000000)/100000)

        [{st_d, st_d, xx1,xx2,xx3,xx4,xx5,xx6}] ++
        [{en_d, en_d, xx1,xx2,xx3,xx4,xx5,xx6}]
      true ->
        br({st_d, t_half, x1,x2,x3,x4,x5,x6}) ++
        br({t_half, en_d, x1,x2,x3,x4,x5,x6})
    end
  end

  def start() do
    pid_res_srv = spawn(__MODULE__, :cnt, [])
    :global.re_register_name('res_srv', pid_res_srv)
    Logger.info("World.start res PID=#{inspect(pid_res_srv)}")

    World.start(__MODULE__, [{0,999999, 0,0,0,0,0,0}])
    :ok
  end
end
